JDK-8346465 : Add a check in setData() to restrict the update of Built-In ICC_Profiles

honkar-jdk opened the pull request:
Built-in Profiles are singleton objects and if the user happens to modify this shared profile object via setData() then the modified version of the profile is returned each time the same built-in profile is requested via getInstance().

It is good to protect Built-in profiles from such direct modification by adding BuiltIn profile check in `setData()` such that **only copies** of Built-In profiles are allowed to be updated.

With the proposed fix, if Built-In profile is updated using `.setData()` it throws _**IAE - "BuiltIn profile cannot be modified"**_

Fix consists of:
* `private boolean isBuiltIn = false` in ICC_Profile which is set to true when BuiltIn profiles are created and false for non-builtIn profile.
* Converted BuiltInProfile from private interface to private static class (with all the profile instances as static final)
* `isBuiltIn` flag is set to true when BuiltInProfile are loaded.
* JavaDoc update for setData() (CSR required)

There are no restrictions on creating copies of BuiltIn profile and then modifying it, but what is being restricted with this fix is - the direct modification of the shared BuiltIn profile instance.

Applications which need a modified version of the ICC Profile should instead do the following:

```
byte[] profileData = ICC_Profile.getData() // get the byte array represtation of BuiltIn- profile
ICCProfile newProfile = ICC_Profile.getInstance(profileData) // create a new profile
newProfile.setData() // to modify and customize the profile
```

Following existing tests are modified to update a copy of Built-In profile.

- java/awt/color/ICC_Profile/SetHeaderInfo.java
- java/awt/color/ICC_ProfileSetNullDataTest.java
- sun/java2d/cmm/ProfileOp/SetDataTest.java

NOTE : For existing tests only necessary updates are done.

honkar-jdk commented:
/csr needed

prrace commented on src/java.desktop/share/classes/java/awt/color/ICC_Profile.java:
```diff
@@ -1139,9 +1161,16 @@ private static byte[] getData(Profile p, int tagSignature) {
      *         array can not be interpreted as valid tag data, corresponding to
      *         the {@code tagSignature}
+     * @throws IllegalArgumentException if the profile being updated has
+     *         {@code isBuiltIn} flag set to true
```
isBuiltin is a private variable. You can't reference it in public API javadoc.

What this ought to say is
@throw IllegalArgumentException if this is a profile for one of the built-in pre-defined ColorSpaces, i.e those which can be obtained by calling ICC_Profile.getInstance(int cspace) 

There isn't a general way to get that list - but it is short - only 5 colour spaces, and you could put an @see ColorSpace link

honkar-jdk replied:
Thanks Phil. Updated.

honkar-jdk commented on src/java.desktop/share/classes/java/awt/color/ICC_Profile.java:
```diff
@@ -1129,7 +1150,8 @@ private static byte[] getData(Profile p, int tagSignature) {
      * {@code tagSignature} as defined in the ICC specification, section 10.
      * This method is useful for advanced applications which need to access
-     * profile data directly.
+     * profile data directly. Only non-BuiltIn profiles can be updated using
+     * this method.
```
@prrace 
I added this additional line here mentioning that only non-BuiltIn profiles can be updated. Does it look okay or do you recommend any changes ?

jayathirthrao replied:
I think here also language should be similar to what we use in `@throws IllegalArgumentException`. Using just "non-BuiltIn profiles" doesn't give full picture.

honkar-jdk replied:
@prrace @jayathirthrao  Thanks! Updated both the PR as per CSR changes.

prrace commented on src/java.desktop/share/classes/java/awt/color/ICC_Profile.java:
```diff
@@ -1139,9 +1161,18 @@ private static byte[] getData(Profile p, int tagSignature) {
      *         the {@code tagSignature}
+     * @throws IllegalArgumentException if this is a profile for one of the
+     *         built-in pre-defined ColorSpaces, i.e. those which can be obtained
+     *         by calling ICC_Profile.getInstance(int cspace)
```
you should put that method ref in {@code .. }

aivanov-jdk commented on src/java.desktop/share/classes/java/awt/color/ICC_Profile.java:
```diff
@@ -1129,7 +1150,8 @@ private static byte[] getData(Profile p, int tagSignature) {
      * {@code tagSignature} as defined in the ICC specification, section 10.
      * This method is useful for advanced applications which need to access
-     * profile data directly.
+     * profile data directly. Only non-built-in, application provided profiles
+     * should be updated using this method.
```
What you mean is that only non-built-in profiles *can* be updated. The fix makes it impossible to update built-in profiles.

honkar-jdk replied:
I'll be inverting this statement to state what is NOT allowed rather than what is allowed by .setData() as below. Does the following sound better?

```
     * Note: JDK built-in ICC Profiles cannot be updated using this method
     * as it will result in IAE. JDK built-in profiles are those obtained by
     * {@code ICC_Profile.getInstance(int colorSpaceID)} where colorSpaceID
     * is one of the following:
     * {@link ColorSpace#CS_sRGB}, {@link ColorSpace#CS_LINEAR_RGB},
     * {@link ColorSpace#CS_PYCC}, {@link ColorSpace#CS_GRAY} or
     * {@link ColorSpace#CS_CIEXYZ}
```

aivanov-jdk replied:
This is clearer.

honkar-jdk replied:
Looks like my previous commit was not reflected. I have update it now.

@prrace Updated PR and CSR.

aivanov-jdk commented on src/java.desktop/share/classes/java/awt/color/ICC_Profile.java:
```diff
@@ -1139,9 +1161,18 @@ private static byte[] getData(Profile p, int tagSignature) {
      * @throws IllegalArgumentException if a content of the {@code tagData}
      *         array can not be interpreted as valid tag data, corresponding to
      *         the {@code tagSignature}
+     * @throws IllegalArgumentException if this is a profile for one of the
```
`IllegalStateException` better describes the reason: the argument to the method can be perfectly valid, but the internal state of the object doesn't allow modifications.

honkar-jdk replied:
@aivanov-jdk 

_IllegalStateException  - Signals that a method has been invoked at an **illegal or inappropriate time.**_  
Since IllegalStateException is thrown to indicate more of an unstable state of object, it may not be what we want here. The exception is to be thrown when the ICC_Profile object invoking .setData() is JDK Built-in profile and IIlegalArgumentException more closely match this case.

aivanov-jdk replied:
The javadoc for `IllegalArgumentException` says, <q cite="https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/IllegalArgumentException.html">Thrown to indicate that a method has been passed <em>an illegal or inappropriate argument</em>.</q> (Emphasis mine.)

Getting an `IllegalArgumentException` for a valid argument would be confusing.

`IllegalStateException`, as you said, “signals that a method has been invoked at an <em>illegal or inappropriate</em> time”. It's exactly the state of the object: a *built-in* color profile cannot be modified.

Although, there's no time where such a modification will be allowed, I still think `IllegalStateException` better conveys the meaning: changes to built-in profiles are *never* allowed.

honkar-jdk replied:
> Although, there's no time where such a modification will be allowed

Exactly, it will throw exception every time a built-In profile is passed unlike IllegalStateException which is thrown only during an inappropriate time for instance when you are trying to add an element to the queue when it is already full. Plus setData() already throws IAE for other invalid argument cases.

prrace replied:
Although it is not a checked exception, the fact that IllegalArgumentException is already documented on this method is considered to outweigh any benefit of a new Exception type such as IllegalStateException. And IllegalArgumentException is not that far off from existing usage.
Existing usage includes the case that "the specified data is not appropriate to be set in this profile".

aivanov-jdk replied:
> > Although, there's no time where such a modification will be allowed
> 
> Exactly, it will throw exception every time a built-In profile is passed unlike IllegalStateException which is thrown only during an inappropriate time for instance when you are trying to add an element to the queue when it is already full. Plus setData() already throws IAE for other invalid argument cases.

I see no contradiction here. Inappropriate time is a vague definition. For built-in object, it is always inappropriate time to call `setData`.

`IllegalArgumentException` implies that the argument is invalid, and if I change the argument, the call will succeed — but the call will never succeed for a built-in profile *because **the state of the object** doesn't allow it*.

This is why `IllegalStateException` is more appropriate.

> Although it is not a checked exception, the fact that IllegalArgumentException is already documented on this method is considered to outweigh any benefit of a new Exception type such as IllegalStateException.

I do not agree here. The exception type should convey the reason it's thrown, otherwise there wouldn't be so many types of exceptions.

And two `@throws` with the same type aren't allowed in javadoc, which means you have to add the new reason to throw `IllegalArgumentException` into the existing one.

aivanov-jdk replied:
Let's consider a scenario.

If I call `setData` with `tagSignature` and `tagData` which cannot be interpreted correctly, I get `IllegalArgumentException`, which is reasonable — I passed invalid arguments.

Then if I call `setData` with valid values for `tagSignature` and `tagData`, the call succeeds even with a built-in color profile. After this fix, this call would fail with `IllegalArgumentException` but I know that the arguments are **valid** because the call succeeded previously. And however I change the arguments, the call will never succeed.

Throwing `IllegalStateException` will convey the updated behaviour in a cleaner way: it is *the state* of the object that makes the call to fail, not the arguments.

prrace replied:
> And two @throws with the same type aren't allowed in javadoc, which means you have to add the new reason to throw IllegalArgumentException into the existing one.

That's demonstrably untrue. This method already has two.

Regarding the scenario above, this apparent and has all been considered already and the IAE is the preferred solution.

And FWIW I don't think ISE is really any better.

aivanov-jdk replied:
> > And two @throws with the same type aren't allowed in javadoc, which means you have to add the new reason to throw IllegalArgumentException into the existing one.
> 
> That's demonstrably untrue. This method already has two.

I was wrong here, the specification for [`ICC_Profile.html.setData`](https://docs.oracle.com/en/java/javase/23/docs/api/java.desktop/java/awt/color/ICC_Profile.html#setData(int,byte[])) correctly displays the two entries for `IllegalArgumentException`. The generated javadoc for the proposed changeset contains three entries for IAE.

I misremembered it, or it was a limitation in an IDE javadoc parser.

aivanov-jdk replied:
> Regarding the scenario above, this apparent and has all been considered already and the IAE is the preferred solution.
> 
> And FWIW I don't think ISE is really any better.

<code>Illegal<i>Argument</i>Exception</code> implies the *argument* is invalid, but it is now thrown for a *valid* argument if the object *state* doesn't allow modifying the data.

<code>Illegal<i>State</i>Exception</code> conveys the object *state* is inappropriate to call this method.

There's a semantic difference between the two, and I think <code>Illegal<i>State</i>Exception</code> is better in this case.

We can also discuss it with Joe in the CSR.

honkar-jdk replied:
@aivanov-jdk  As Phil mentioned earlier the already documented exception IAE for `.setData()` outweighs introducing a new exception. Using an existing exception would mean less disruption to any existing applications using the Java API.

aivanov-jdk commented on src/java.desktop/share/classes/java/awt/color/ICC_Profile.java:
```diff
@@ -107,30 +107,51 @@ public sealed class ICC_Profile implements Serializable
+     * This check is used in {@link #setData(int, byte[])} to prevent modifying
+     * BuiltInProfile.
+     */
+    private boolean isBuiltIn = false;
```
Indeed, it *can be*, and I prefer this design. This way, the `isBuiltIn` field can't be changed after the object is constructed.

To achieve this, you have to add a constructor which accepts a boolean flag: `ICC_Profile(ProfileDeferralInfo pdi, boolean builtIn)`, and `ICC_ProfileRGB` and `ICC_ProfileGray` need to provide an additional constructor as well.

Here's [a diff to your branch which achieves this](https://github.com/honkar-jdk/jdk/compare/BuiltInCheck...aivanov-jdk:jdk:harshitha/8346465-built-inprofiles).

When [you compare the changeset that I propose to `jdk/master`](https://github.com/openjdk/jdk/compare/master...aivanov-jdk:jdk:harshitha/8346465-built-inprofiles), there are less differences, and `BuiltInProfile` remains an interface. Yet two more files `ICC_ProfileRGB` and `ICC_ProfileGray` are modified.

aivanov-jdk replied:
After adding a constructor which accepts a boolean parameter, the constructors `ICC_ProfileRGB(ProfileDeferralInfo pdi)` and `ICC_ProfileGray(ProfileDeferralInfo pdi)` — these two can be removed then.

honkar-jdk replied:
Both the approaches - setting isBuiltIn in constructor vs in the static block, were evaluated previously. Setting it via constructor meant that we could mark the profile as Built-In only when it was constructed using ProfileDeferralInfo and modifying all the constructors - ICC_Profile, ICC_ProfileRGB, ICC_ProfileGray whereas setting it in static block meant minimal changes to existing code.

aivanov-jdk replied:
I don't get the reason why it's bad to modify the constructors.

At this time, all the built-in profiles are created with `ProfileDeferralInfo`.

If `ProfileDeferralInfo` is used to create only built-in profiles, then modifying `ICC_Profile(ProfileDeferralInfo pdi)` is enough:

```java
    ICC_Profile(ProfileDeferralInfo pdi) {
        deferralInfo = pdi;
        isBuiltIn = true;
    }
```

And this is a *minimal* change.

When the `isBuiltIn` is final, it's much easier to find where the value gets set.

aivanov-jdk replied:
> At this time, all the built-in profiles are created with `ProfileDeferralInfo`.
> 
> If `ProfileDeferralInfo` is used to create only built-in profiles, then modifying `ICC_Profile(ProfileDeferralInfo pdi)` is enough.

I've updated my branch with this change: `ICC_Profile(ProfileDeferralInfo pdi)` sets the `isBuiltIn` flag to `true`; `ICC_Profile(Profile p)` sets `isBuiltIn` to `false`.

The diff to ‘master’: https://github.com/openjdk/jdk/compare/2a5d1da3355a4df3109ec42646b5b0cf088b4c2a..a34f16860c2f7d393f4a1fb57f46fba1a68a8412

Only `ICC_Profile.java` is modified, which is *minimal*. Only several lines are added.

If there's another way to create a built-in profile in the future, this approach will need to updated to take into account the new way. We'll deal with it at that time in the future.

The diff to your branch: https://github.com/openjdk/jdk/compare/8da82c10b5be3b92655abef95fd5be0c8b6eaa00..a34f16860c2f7d393f4a1fb57f46fba1a68a8412

honkar-jdk replied:
There are other way to create a profile. Directly loading it through a file. 
`ICC_Profile.getInstance("/sRGB.pf"); `or using just the byte array representation of the profile. So the main intention here was not to tie ProfileDeferralInfo with isBuiltIn.

aivanov-jdk replied:
> There are other way to create a profile - directly loading it form a file (serialization) `ICC_Profile.getInstance(<path to sRGB.pf or any custom profiles>); `or using the byte array representation of the profile. So the main intention here was not to tie ProfileDeferralInfo with isBuiltIn.

Yes, there are. Does any other way create a **built-in profile**? No, it doesn't as far as I can see.

Is this flexibility needed? I'd say, it's not needed… unless there's a very high chance there'll soon be introduced a new build-in ICC profile which is created in another way but `ICC_Profile(ProfileDeferralInfo)` constructor.

honkar-jdk replied:
> Yes, there are. Does any other way create a built-in profile? No, it doesn't as far as I can see.

Yes, currently ProfileDeferralInfo  is the only way to create built-in profile.

> Is this flexibility needed? I'd say, it's not needed… unless there's a very high chance there'll soon be introduced a new build-in ICC profile which is created in another way but ICC_Profile(ProfileDeferralInfo) constructor.

As of now we don't need the flexibility but not sure if that is something that we may require r need to consider in future. @prrace Your suggestion ?

prrace replied:
So in fact Harshitha had prototyped the constructor mechanism - but it was passing a parameter so wasn't as small a diff as the most recent alternate, because I think it rippled into changing the sub-class constructors as well, which made it less desirable.
Another particular concern was that it was equating ProfileDeferralInfo to meaning isBuiltIn == true, which is the case today, but it is not a fundamental truth.
And the proposal here is more direct and you don't have to go follow the bread crumbs to see exactly which profiles have this set to true.  So that is why the PR is as it is.
Neither is "wrong" - they are just different choices based on different perceptions.

But if we can do this with changes to *just this class*, then may be .. and since I see the sub-class constructors (for RGB and Gray) call the same signature super-class constructors, then I think this should work, and it will be OK.
I tried it on a local repo, and came up with the same changes as suggested and it least builds .. 
So we can do it that way too, but there should be a comment somewhere that ProfileDeferralInfo is only for built-in profiles - and that in fact all built-in profiles should use it.

honkar-jdk replied:
Thanks for the update on this discussion Phil. I'll update the code to set built-in flag in the constructor and add a comment stating - ProfileDeferralInfo to be used only for built-in profiles.

honkar-jdk replied:
@aivanov-jdk Updated the code and moved `builtIn` flag to be set in the constructor.

aivanov-jdk commented on src/java.desktop/share/classes/java/awt/color/ICC_Profile.java:
```diff
@@ -1131,6 +1152,16 @@ private static byte[] getData(Profile p, int tagSignature) {
      * This method is useful for advanced applications which need to access
      * profile data directly.
      *
+     * <p>
+     * Note: JDK built-in ICC Profiles cannot be updated using this method
+     * as it will result in IAE. JDK built-in profiles are those obtained by
+     * {@code ICC_Profile.getInstance(int colorSpaceID)} where colorSpaceID
+     * is one of the following:
+     * {@link ColorSpace#CS_sRGB}, {@link ColorSpace#CS_LINEAR_RGB},
+     * {@link ColorSpace#CS_PYCC}, {@link ColorSpace#CS_GRAY} or
+     * {@link ColorSpace#CS_CIEXYZ}.
+     * </p>
```
```suggestion
     * <p>
     * Note: JDK built-in ICC Profiles cannot be updated using this method
     * as it will result in {@code IllegalArgumentException}. JDK built-in
     * profiles are those obtained by
     * {@code ICC_Profile.getInstance(int colorSpaceID)}
     * where {@code colorSpaceID} is one of the following:
     * {@link ColorSpace#CS_sRGB}, {@link ColorSpace#CS_LINEAR_RGB},
     * {@link ColorSpace#CS_PYCC}, {@link ColorSpace#CS_GRAY} or
     * {@link ColorSpace#CS_CIEXYZ}.
     * </p>
```
You should spell `IllegalArgumentException` fully, it explicitly lists the exception thrown. And `colorSpaceID` should also be marked up with `{@code}`.

aivanov-jdk commented on src/java.desktop/share/classes/java/awt/color/ICC_Profile.java:
```diff
@@ -1139,9 +1170,18 @@ private static byte[] getData(Profile p, int tagSignature) {
      * @throws IllegalArgumentException if a content of the {@code tagData}
```
This is an existing issue, if it's an issue: should the text say *“**the** content”* instead of *“**a** content”*? @prrace

prrace replied:
Yes.

honkar-jdk replied:
Updated.

aivanov-jdk commented on src/java.desktop/share/classes/java/awt/color/ICC_Profile.java:
```diff
@@ -1139,9 +1170,18 @@ private static byte[] getData(Profile p, int tagSignature) {
      * @throws IllegalArgumentException if a content of the {@code tagData}
      *         array can not be interpreted as valid tag data, corresponding to
      *         the {@code tagSignature}
+     * @throws IllegalArgumentException if this is a profile for one of the
+     *         built-in pre-defined ColorSpaces, i.e. those which can be obtained
+     *         by calling {@code ICC_Profile.getInstance(int colorSpaceID)}
```
```suggestion
     * @throws IllegalArgumentException if this is a built-in profile for one of the
     *         pre-defined color spaces, i.e. those which can be obtained
     *         by calling {@code ICC_Profile.getInstance(int colorSpaceID)}
```

aivanov-jdk commented on test/jdk/java/awt/color/ICC_Profile/BuiltInProfileCheck.java:
```diff
@@ -0,0 +1,80 @@
+
+    private static void updateProfile(boolean isBuiltIn) {
+        ICC_Profile builtInProfile = ICC_Profile.getInstance(ColorSpace.CS_sRGB);
+        //create a copy of the BuiltIn profile
```
```suggestion
        // Create a copy of the built-in profile
```

aivanov-jdk commented on test/jdk/java/awt/color/ICC_Profile/BuiltInProfileCheck.java:
```diff
@@ -0,0 +1,80 @@
+
+        if (isBuiltIn) {
+            try {
+                //try updating a BuiltIn Profile, the following stmt should throw IAE
```
```suggestion
                // Try updating a built-in profile, IAE is expected
```

aivanov-jdk commented on test/jdk/java/awt/color/ICC_Profile/BuiltInProfileCheck.java:
```diff
@@ -0,0 +1,80 @@
+            try {
+                //try updating a BuiltIn Profile, the following stmt should throw IAE
+                iccProfile.setData(HEADER_TAG, headerData);
+                throw new RuntimeException("Test Failed ! IAE NOT thrown.");
```
```suggestion
                throw new RuntimeException("Test Failed! IAE NOT thrown.");
```

aivanov-jdk commented on test/jdk/java/awt/color/ICC_Profile/BuiltInProfileCheck.java:
```diff
@@ -0,0 +1,80 @@
+
+        byte[] headerData = iccProfile.getData(HEADER_TAG);
+        int index = PROFILE_CLASS_START_INDEX;
+        //set profile class to valid icSigInputClass = 0x73636E72
```
```suggestion
        // Set profile class to valid icSigInputClass = 0x73636E72
```

aivanov-jdk commented on test/jdk/java/awt/color/ICC_Profile/BuiltInProfileCheck.java:
```diff
@@ -0,0 +1,80 @@
+                }
+            }
+        } else {
+            //modifying custom profile should NOT throw IAE
```
```suggestion
            // Modifying custom profile should NOT throw IAE
```

aivanov-jdk commented on test/jdk/java/awt/color/ICC_ProfileSetNullDataTest.java:
```diff
@@ -30,22 +30,23 @@
  * @summary Test checks behavior of the ICC_Profile.setData(int, byte[])
  */
 public final class ICC_ProfileSetNullDataTest {
+    private static final int[] colorSpace = new int [] {
```
```suggestion
    private static final int[] colorSpace = new int[] {
```

aivanov-jdk replied:
In fact, the syntax can be simplified further:
```java
private static final int[] colorSpace = {
```

There's no need for explicit `new int[]`.

aivanov-jdk commented on test/jdk/java/awt/color/ICC_ProfileSetNullDataTest.java:
```diff
@@ -30,22 +30,23 @@
+            ICC_Profile profile = ICC_Profile.getInstance(builtInProfile.getData());
+            try {
+                profile.setData(ICC_Profile.icSigCmykData, tagData);
+            } catch (IllegalArgumentException e) {
+                return;
+            }
```
The test won't run for **all** the cases if you after the first successful case — use `continue`.

honkar-jdk replied:
Good catch. Moved ` throw new RuntimeException("IllegalArgumentException expected");`  within try block so continue no longer required.

aivanov-jdk commented on test/jdk/java/awt/color/ICC_Profile/BuiltInProfileCheck.java:
```diff
@@ -0,0 +1,80 @@
+    public static void main(String[] args) {
+        System.out.println("CASE 1: Testing BuiltIn Profile");
+        updateProfile(true);
+        System.out.println("Passed \n");
```
```suggestion
        System.out.println("Passed\n");
```

aivanov-jdk commented on src/java.desktop/share/classes/java/awt/color/ICC_Profile.java:
```diff
@@ -107,30 +107,51 @@ public sealed class ICC_Profile implements Serializable
+     * This check is used in {@link #setData(int, byte[])} to prevent modifying
+     * BuiltInProfile.
+     */
+    private boolean isBuiltIn = false;
```
Should the field be named `builtIn` instead? `isBuiltIn` starts with a verb which is used for naming methods.

honkar-jdk replied:
IIRC we usually have `is` or `has` prefixed to boolean vars as per naming convention?

honkar-jdk replied:
You are right, probably fields are named without the prefix and methods/local vars are prefixed with 'is'.

aivanov-jdk commented on test/jdk/java/awt/color/ICC_Profile/BuiltInProfileCheck.java:
```diff
@@ -48,32 +48,32 @@ public static void main(String[] args) {
             } catch (IllegalArgumentException iae) {
-                if (!iae.getMessage().equals(EXCEPTION_MSG)) {
-                    throw new RuntimeException("Test Failed ! IAE with exception msg \""
+                if (!iae.getMessage().equalsIgnoreCase(EXCEPTION_MSG)) {
```
I think we can compare with regular `equals`. It is unlikely the message will change by letter case only. The test uses exactly the same message as the one that's thrown in `ICC_Profile.setData`, therefore the messages should match.

If the message is changed in the code for whatever reason, the message in the test will need updating, which is expected, isn't it?

aivanov-jdk commented:
@mrserb You may want to take a look.

mrserb commented on src/java.desktop/share/classes/java/awt/color/ICC_Profile.java:
```diff
@@ -118,19 +126,19 @@ private interface BuiltInProfile {
 
         ICC_Profile PYCC = new ICC_Profile(new ProfileDeferralInfo(
-               "PYCC.pf", ColorSpace.TYPE_3CLR, 3, CLASS_COLORSPACECONVERSION));
+                "PYCC.pf", ColorSpace.TYPE_3CLR, 3, CLASS_COLORSPACECONVERSION));
```
The change in this block will exceed 80 characters per line.

mrserb commented on src/java.desktop/share/classes/java/awt/color/ICC_Profile.java:
```diff
@@ -763,14 +771,21 @@ private interface BuiltInProfile {
+     * <p>
+     * Note: {@code ProfileDeferralInfo} is used for only built-in profile
+     * creation and all built-in profiles should be constructed using it.
+     * </p>
```
</p> is not necessary here.

honkar-jdk replied:
This note was added based on the suggestion here: https://github.com/openjdk/jdk/pull/23606#discussion_r1976123437

honkar-jdk replied:
Oh you meant `<p>` tag :) 
It was not rendered in github comment.

mrserb commented on src/java.desktop/share/classes/java/awt/color/ICC_Profile.java:
```diff
@@ -1131,17 +1146,36 @@ private static byte[] getData(Profile p, int tagSignature) {
+     * where {@code colorSpaceID} is one of the following:
+     * {@link ColorSpace#CS_sRGB}, {@link ColorSpace#CS_LINEAR_RGB},
+     * {@link ColorSpace#CS_PYCC}, {@link ColorSpace#CS_GRAY} or
+     * {@link ColorSpace#CS_CIEXYZ}.
```
The standard profile could be acquired by the application via serialization as well, but I'm not sure if we should mention it.

honkar-jdk replied:
This builtin flag check is exclusively for JDK built-in profiles created within the private interface `BuiltInProfile` and not applicable if an application loads the profile by serialization. 

Since `BuiltInProfile` interface and `builtin` flag are private we could not include it in javadoc hence used ICC_Profile.getInstance(int colorspaceID) to specify built-in profile. Please let us know if the javadoc sound okay or requires any changes?

mrserb replied:
> This builtin flag check is exclusively for JDK built-in profiles created within the private interface `BuiltInProfile` and not applicable if an application loads the profile by serialization.

The built-in profiles can be serialized and deserialized, [see](https://github.com/openjdk/jdk/blob/c4b516dfe7c5a5fddd4d9c97a21f5f36bf845646/src/java.desktop/share/classes/java/awt/color/ICC_Profile.java#L1573). This means the application may get a reference to the profile not only via the getInstance(id) method. Therefore, the new spec is not strictly accurate regarding ICC_Profile.getInstance(int colorSpaceID), but I am not sure that we should mention serialisation here.

**Note:** 
The new Javadoc for the new private built-in flag is "public" via the serialized form of the class. It probably should be marked as transient, so it will only be set to true when the built-in profile is loaded via serialization. In all other cases, it will be false.

aivanov-jdk replied:
This is an interesting observation.

Are the unserialized built-in profiles mapped to their corresponding singleton instances? I presume it's the case. Then, the `builtIn` flag will still be set to `true` in the returned object.

**Should the serialized form be preserved?**

If the `builtIn` field is `transient`, built-in profiles written by JDK 25 (mainline) could still be read in previous versions and vice versa. This will improve the backward compatibility.

honkar-jdk replied:
Good point. I had not considered the effect of serialization on builtIn flag, let me check and get back.

prrace replied:
Yes, readObject(..) and readResolve(..) will return the singleton, as per the serialization spec 👍 https://docs.oracle.com/en/java/javase/21/docs/api/serialized-form.html#java.awt.color.ICC_Profile
but I agree we don't need to mention that here.

Yes, I agree builtIn should be transient, since we return the singleton as above, all will work as expected.

mrserb replied:
>Yes, readObject(..) and readResolve(..) will return the singleton

We do not specify anywhere in the API that we use singletons; that is an implementation detail.

honkar-jdk replied:
builtIn flag converted into transient.

mrserb commented on src/java.desktop/share/classes/java/awt/color/ICC_Profile.java:
```diff
@@ -1131,17 +1146,36 @@ private static byte[] getData(Profile p, int tagSignature) {
      *         the {@code tagSignature}
+     * @throws IllegalArgumentException if this is a built-in profile for one
+     *         of the pre-defined ColorSpaces, i.e. those which can be obtained
+     *         by calling {@code ICC_Profile.getInstance(int colorSpaceID)}
```
can we just ignore it instead and did not use suspicion IllegalArgumentException for correct parameters? or change the type to something unrelated to "..ArgumentException"?

prrace replied:
The silently do nothing option was considered, but if you do that, then you have no easy way of knowing if it worked.
Tests may pass spuriously, or fail later for the wrong reasons. So a worse choice.
And in all my searching of uses of this API it is (1) tests in the JDK itself and (2) a couple of libraries that are targeting specific known profiles with issues and are fixed up - so never applied to built-in profiles.

aivanov-jdk replied:
> can we just ignore it instead and did not use suspicion IllegalArgumentException for correct parameters? or change the type to something unrelated to "..ArgumentException"?

This has been [my suggestion](https://github.com/openjdk/jdk/pull/23606#discussion_r1968073992) since the start of this code review.

> The silently do nothing option was considered, but if you do that, then you have no easy way of knowing if it worked.  
> Tests may pass spuriously, or fail later for the wrong reasons. So a worse choice.

I agree, an exception should be thrown if modification isn't allowed.

> And in all my searching of uses of this API it is (1) tests in the JDK itself and (2) a couple of libraries that are targeting specific known profiles with issues and are fixed up - so never applied to built-in profiles.

Therefore we can throw <code>Illegal<i>State</i>Exception</code> instead of <code>Illegal<i>Argument</i>Exception</code>.

Quoting here [Harshitha's reply](https://github.com/openjdk/jdk/pull/23606#discussion_r1968263836):

> As Phil mentioned earlier, the already documented exception IAE for .setData() outweighs introducing a new exception.

Using the exception which clearly describes the reason why it was thrown outweighs re-using an already documented exception…

> Using an existing exception would mean less disruption to any existing applications using the Java API.

…especially because the search for usages of `setData` revealed that there are no usages of `setData` which will be affected by the proposed change.

If I'm not mistaken, the search also showed that no usages of `setData` handle `IllegalArgumentException` — thus apps will be equally unprepared to any type of an unchecked exception.

honkar-jdk replied:
@aivanov-jdk @mrserb 
There are two other RTE types that may be better suited that ISE in this case - [UnsupportedOperationException ](https://docs.oracle.com/en/java/javase/23/docs/api/java.base/java/lang/UnsupportedOperationException.html) and [ProfileDataException](https://docs.oracle.com/en/java/javase/21/docs/api/java.desktop/java/awt/color/ProfileDataException.html) but the main motivation for going with IAE is that it is existing and documented exception.

> If I'm not mistaken, the search also showed that no usages of setData handle IllegalArgumentException — thus apps will be equally unprepared to any type of an unchecked exception.

Reasonable point.

mrserb replied:
I'm just not sure how this should be handled from the application's point of view. The app may obtain a profile from some place and use it for some images or pixels. Then, if the app wants to tweak the rendering intent for some reason, what should it do?

 - Clone the profile and then change the intent?
 - Change the intent, and if an exception occurs, clone the profile and try to change the intent again?
 - Skip changing the intent completely?

I remember we discussed this issue during the review of [this commit](https://github.com/openjdk/jdk/pull/3037/files#diff-3dd53a33889801159f43dbb990ba033066bdabaed71bbc7254f58331d3898d69R39), and the conclusion was that it was too late to change it and break apps. It wasn’t considered very important since the "appcontext was withdrawn". So, why are we deciding to change it now?

prrace replied:
> I'm just not sure how this should be handled from the application's point of view. The app may obtain a profile from some place and use it for some images or pixels. Then, if the app wants to tweak the rendering intent for some reason, what should it do?
> 
> * Clone the profile and then change the intent?
> * Change the intent, and if an exception occurs, clone the profile and try to change the intent again?
> * Skip changing the intent completely?

If an app got a profile 'from some place', it probably should not be trying to change it, and likely has no idea if some isolated change will work properly, or the effect on what handed it that profile.
So copying the data and creating a new profile is what I'd recommend in such a case, regardless of this proposed restriction, and even then, only on a 'known' profile. 

> I remember we discussed this issue during the review of [this commit](https://github.com/openjdk/jdk/pull/3037/files#diff-3dd53a33889801159f43dbb990ba033066bdabaed71bbc7254f58331d3898d69R39), and the conclusion was that it was too late to change it and break apps. It wasn’t considered very important since the "appcontext was withdrawn". So, why are we deciding to change it now?

I don't see such a discussion there. Maybe it was 'offline' ?
Yes, this is less of an issue in a closed application, but it is an inappropriate thing to do/allow.

mrserb replied:
>If an app got a profile 'from some place', it probably should not be trying to change it,

But this is not always possible. For example, a buffered image has a color model that references a color space and a color profile. To modify something in that profile(like intents screen vs printer) before converting the entire image, it is necessary to check whether it is a built-in profile. If it is, the whole image must be recreated; otherwise, an exception will occur. Since there is no way to check if a profile is built-in, making built-in profiles read-only might be a better approach, similar to [java properties](https://bugs.openjdk.org/browse/JDK-8066709).

>I don't see such a discussion there. Maybe it was 'offline' ?

It was somewhere online not too long ago.

mrserb replied:
>Since there is no way to check if a profile is built-in, making built-in profiles read-only might be a better approach, similar to [java properties](https://bugs.openjdk.org/browse/JDK-8066709).

But it also has its own disadvantages...

prrace replied:
The System Properties case isn't something I'd want to follow. It doesn't really make them read-only.
It just ensures that the JDK internally caches the original value for use in certain places.

Consider this program

public class jh {
  public static void main(String[] args) {
   String jhpstr = "java.home";
   String jhp1 = System.getProperty(jhpstr);
   System.setProperty(jhpstr, "foo");
   String jhp2 = System.getProperty("java.home");
   System.out.println(jhp1+" " +jhp2);
  }
} 
% java jh
/Library/Java/JavaVirtualMachines/jdk-21.jdk/Contents/Home foo

aivanov-jdk replied:
> There are two other exceptions that may be better suited that ISE in this case - [UnsupportedOperationException ](https://docs.oracle.com/en/java/javase/23/docs/api/java.base/java/lang/UnsupportedOperationException.html) and [ProfileDataException](https://docs.oracle.com/en/java/javase/21/docs/api/java.desktop/java/awt/color/ProfileDataException.html) but still IAE was chosen over them as it is existing and documented exception.

`UnsupportedOperationException` could be used.

`HeadlessException` extends `UnsupportedOperationException` and indicates a situation where a method cannot proceed because of the platform configuration.

`ProfileDataException` doesn't suit either because there's nothing wrong with the profile data in the case of built-in profiles.

mrserb replied:
> The System Properties case isn't something I'd want to follow. It doesn't really make them read-only.
> It just ensures that the JDK internally caches the original value for use in certain places.

I think you missed the main point. The application may attempt to change these java properties, but it won't affect the behavior of the JDK since the JDK internally will use the initially cached value of the property via StaticProperty.userDir()/userHome()/etc.

This is even specified similarly to what we want for color profiles.

>  * <strong>Changing a standard system property may have unpredictable results
     * unless otherwise specified.</strong>
     * Property values may be cached during initialization or on first use.
     * Setting a standard property after initialization using {@link #getProperties()},
     * {@link #setProperties(Properties)}, {@link #setProperty(String, String)}, or
     * {@link #clearProperty(String)} may not have the desired effect.

mrserb replied:
* Builtin color profiles may be cached during initialization or on first use. Changing a standard profiles after initialization using ICC_Profile.setData may not have the desired effect.

honkar-jdk replied:
> > Since there is no way to check if a profile is built-in, making built-in profiles read-only might be a better approach, similar to [java properties](https://bugs.openjdk.org/browse/JDK-8066709).
> 
> But it also has its own disadvantages...

@mrserb  You mentioned earlier that there are disadvantages of using java properties approach, so using builtIn flag would be better then, correct?

>  I remember we discussed this issue during the review of [this commit](https://github.com/openjdk/jdk/pull/3037/files#diff-3dd53a33889801159f43dbb990ba033066bdabaed71bbc7254f58331d3898d69R39), and the conclusion was that it was too late to change it and break apps.

I wasn't able to find the previous discussion thread so I'm not sure what are the disadvantages. Can you please explain?

mrserb replied:
If we start from the beginning. The problem we are trying to solve is preventing changes to our standard profiles so that we can be sure one part of the application does not unexpectedly affect another. This could be achieved in two ways:

- Throw an exception if a standard profile is modified
- Ignore the requested change and use the default profile properties as is

For new applications, we could require always cloning the color profile before modification (via the updated specification); otherwise, the application must always check whether the profile is built-in or not.

However, for older applications, these two solutions lead to different outcomes:

- Throwing the new exception will most likely break the application
- Ignoring the data change allows the application to continue working, but the color conversion result may not be fully accurate

The next question is: how big of an issue is this inaccuracy? For example, if we have an sRGB profile that rejects some modifications, we would still try to produce valid sRGB data during conversion.

So the trade-off is: broken applications vs. inaccurate (but still ICC-spec-compliant) conversions.

PS: By the way, why do we want to enforce this rule only for built-in profiles? Why can't a library create similar profiles and mark them as "read-only"?

honkar-jdk replied:
@mrserb 

> The app may obtain a profile from some place and use it for some images or pixels. Then, if the app wants to tweak the rendering intent for some reason, what should it do?
> 
> Clone the profile and then change the intent?

The app would follow the above approach.
There is only one step extra with this fix - creating a copy of built-in profile and then modifying it.

```
byte[] builtInData = ICC_Profile.getInstance(ColorSpace.CS_sRGB).getData(); // get the byte array representation of BuiltIn- profile
ICCProfile newProfile = ICC_Profile.getInstance(builtInData) // create a new profile
newProfile.setData(...)
```


> However, for older applications, these two solutions lead to different outcomes:
> Throwing the new exception will most likely break the application

The chances of breaking an existing application is minimal since API usage for setData() was checked and not many were found and moreover they were not called on built-in profiles.


>Ignoring the data change allows the application to continue working, but the color conversion result may not be fully accurate

This does not seem viable because the user in unaware as to whether setData() worked or not.

aivanov-jdk commented on src/java.desktop/share/classes/java/awt/color/ICC_Profile.java:
```diff
@@ -111,9 +111,9 @@ public sealed class ICC_Profile implements Serializable
      * This check is used in {@link #setData(int, byte[])} to prevent modifying
-     * BuiltInProfile.
+     * Built-in profiles.
      */
```
It's recommended to use `{@code}` for keywords, therefore both `true` and `false` should be marked up with `{@code}`.

See [How to Write Doc Comments for the Javadoc Tool, A Style Guide](https://www.oracle.com/technical-resources/articles/java/javadoc-tool.html#styleguide):

> Use <code> style for keywords and names.

prrace commented on test/jdk/java/awt/color/ICC_Profile/BuiltInProfileCheck.java:
```diff
@@ -0,0 +1,80 @@
+                throw new RuntimeException("Test Failed! IAE NOT thrown.");
+            } catch (IllegalArgumentException iae) {
+                if (!iae.getMessage().equals(EXCEPTION_MSG)) {
+                    throw new RuntimeException("Test Failed! IAE with exception msg \""
```
I don't think you need to - or should - expect a specific exception message.
Obviously you can update the test if the text changes, but I just don't think this is necessary.

honkar-jdk replied:
Updated.

honkar-jdk replied:
Reverted based on @aivanov-jdk  suggestion - https://github.com/openjdk/jdk/pull/23606#discussion_r1989287031

mrserb commented on test/jdk/java/awt/color/ICC_ProfileSetNullDataTest.java:
```diff
@@ -30,22 +30,23 @@
+            ColorSpace.CS_sRGB, ColorSpace.CS_LINEAR_RGB,
+            ColorSpace.CS_PYCC, ColorSpace.CS_GRAY,
+            ColorSpace.CS_CIEXYZ
+    };
```
With this modification, we will lose information about which color space is broken from the stack trace. So, at the least, you should print it in case of failure.

honkar-jdk replied:
Updated

aivanov-jdk commented on test/jdk/java/awt/color/ICC_ProfileSetNullDataTest.java:
```diff
@@ -23,27 +23,32 @@
             ICC_Profile builtInProfile = ICC_Profile.getInstance(cs);
             ICC_Profile profile = ICC_Profile.getInstance(builtInProfile.getData());
             try {
                 profile.setData(ICC_Profile.icSigCmykData, tagData);
```
```suggestion
                profile.setData(ICC_Profile.icSigCmykData, null);
```
Should this be simplified?

aivanov-jdk commented on src/java.desktop/share/classes/java/awt/color/ICC_Profile.java:
```diff
@@ -1131,17 +1145,35 @@ private static byte[] getData(Profile p, int tagSignature) {
      */
     public void setData(int tagSignature, byte[] tagData) {
+        if (builtIn) {
+            throw new IllegalArgumentException("Built-in profile"
+                                               + " cannot be modified");
```
Is it acceptable to overshoot 80-column limit for 6 characters without wrapping the line?

aivanov-jdk commented on test/jdk/java/awt/color/ICC_Profile/BuiltInProfileCheck.java:
```diff
@@ -67,10 +66,7 @@ private static void updateProfile(boolean isBuiltIn) {
-                    throw new RuntimeException("Test Failed! IAE with exception msg \""
-                                               + EXCEPTION_MSG + "\" NOT thrown.");
-                }
+                System.out.println("IAE expected: " + iae.getMessage());
```
> test controls the type of arguments (valid arguments)

Yes. And the arguments can change, although it's very unlikely.

The built-in check in the code is done before validating the arguments, therefore we know that whatever exception is thrown, it should be because of trying to modify a built-in profile.

Anyway, checking the exception message seems reasonable, and it ensures the test behaves the way we expect it to.

honkar-jdk commented on src/java.desktop/share/classes/java/awt/color/ICC_Profile.java:
```diff
@@ -763,14 +771,20 @@ private interface BuiltInProfile {
      * Constructs an {@code ICC_Profile} object whose loading will be deferred.
      * The ID will be 0 until the profile is loaded.
+     *
+     * <p>
+     * Note: {@code ProfileDeferralInfo} is used for only built-in profile
+     * creation and all built-in profiles should be constructed using it.
```
Good idea. Added the note near to the location of usage.

honkar-jdk commented on src/java.desktop/share/classes/java/awt/color/ICC_Profile.java:
```diff
@@ -107,6 +107,14 @@ public sealed class ICC_Profile implements Serializable
+     * This check is used in {@link #setData(int, byte[])} to prevent modifying
+     * built-in profiles.
+     */
+    private transient final boolean builtIn;
```
Updated

aivanov-jdk commented on src/java.desktop/share/classes/java/awt/color/ICC_Profile.java:
```diff
@@ -124,6 +124,8 @@ private interface BuiltInProfile {
         /*
          * Deferral is only used for standard profiles. Enabling the appropriate
          * access privileges is handled at a lower level.
```
I wonder if “Enabling the appropriate access privileges is handled at a lower level” is still relevant after `SecurityManager` was removed.

aivanov-jdk replied:
The first sentence can also be removed. The newly added sentence repeats the first one.

aivanov-jdk reviewed (approved):
Looks good to me.

I'm still for using a different exception type rather than `IllegalArgumentException`. The alternatives are `IllegalStateException` or `UnsupportedOperationException`.

mrserb commented on test/jdk/java/awt/color/ICC_Profile/BuiltInProfileCheck.java:
```diff
@@ -0,0 +1,80 @@
+    private static final String EXCEPTION_MSG = "Built-in profile cannot be modified";
+
+    public static void main(String[] args) {
+        System.out.println("CASE 1: Testing BuiltIn Profile");
```
I suggest to add the cases when both(BuiltIn and Custom) profiles are tested after serialisation->deserialisation.

honkar-jdk replied:
Agreed.

honkar-jdk replied:
Updated the test.

mrserb commented on test/jdk/java/awt/color/ICC_Profile/BuiltInProfileCheck.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    private static void updateProfile(boolean isBuiltIn) {
+        ICC_Profile builtInProfile = ICC_Profile.getInstance(ColorSpace.CS_sRGB);
```
It would be good to check all types of BuiltIn profiles to prove that it works in case we decide to implement/change some them differently(w/o cache for example).

honkar-jdk replied:
Updated

mrserb commented on test/jdk/java/awt/color/ICC_Profile/BuiltInProfileCheck.java:
```diff
@@ -0,0 +1,124 @@
+    }
+
+    // this test is added to check the ICC_Profile's transient isBuiltIn flag
+    private static void testSerialization(boolean isBuiltIn)
```
This method does not actually test serialization; instead, it clones/copies the ICC profile via the filesystem. This is why you did not get an exception at the end of this method, even for a built-in profile.

You should use something like this instead:

```java
ByteArrayOutputStream baos = new ByteArrayOutputStream();
ObjectOutputStream oos = new ObjectOutputStream(baos);
oos.writeObject(iccProfile);
```
```java
byte[] array = baos.toByteArray();
ObjectInputStream ois = new ObjectInputStream(new ByteArrayInputStream(array));
ICC_Profile iccProfile = (ICC_Profile) ois.readObject();
```
In the example above, the built-in profile should throw an IAE, but a custom profile should not. (I suggest covering all built-in profiles as well).

mrserb commented:
@prrace I know that the CSR description is not part of the specification. But I would like to point out that we do not specify built-in profiles as singletons (but currently implemented as such). Therefore, it is not necessarily true that the same profile will always be returned by `getInstance` method.

>https://bugs.openjdk.org/browse/JDK-8350025
>The built-in profiles are singleton objects, if an application modifies a shared profile object via ICC_Profile.setData(), then the modified version of the profile is returned each time the same built-in profile is requested via ICC_Profile.getInstance(). This would impact all color space transformations using the modified color space.

prrace commented:
> 

That's an interesting point. The way to obtain them is a get* method which conventionally means it isn't a "new" instance, but the first line of the javadoc unhelpfully says
"Constructs an ICC_Profile".

As a matter of practicality it is hard to imagine that getInstance(int) would return new instances every time in any implementation. Particularly since ICC_Profile doesn't over-ride equals() ! Perhaps we should improve the doc of that method too in some way. But I think that can be a separate PR with its own separate CSR.
The spec. update being made here is valid in either case.

honkar-jdk commented:
@aivanov-jdk As per our discussion I have updated the test code. Please re-review.
Moved profile creation and testing into separate methods - prepareTestProfile(), testProfile()
Added 2 more cases which tests reading .icc files that were serialized on older version of JDK.

@mrserb Can you please re-review the updated test.

honkar-jdk commented on test/jdk/java/awt/color/ICC_Profile/BuiltInProfileCheck/BuiltInProfileCheck.java:
```diff
@@ -0,0 +1,165 @@
+                      : ICC_Profile.getInstance(builtInProfile.getData());
+
+        switch (testCase) {
+            case "BuiltIn", "Custom" -> {
+                // empty case block
+                // no further processing of testProfile required in these 2 cases.
+            }
```
I could move this to a if block before switch and simply return but since switch-case is used I thought to keep all the cases at one place.

aivanov-jdk commented on test/jdk/java/awt/color/ICC_Profile/BuiltInProfileCheck/BuiltInProfileCheck.java:
```diff
@@ -0,0 +1,165 @@
+        System.out.println("Passed\n");
+    }
+
+    private static void prepareTestProfile(String testCase,
+                                           boolean isBuiltIn, int cs) {
```
I'd probably return the resulting profile to run test on, it will make data flow explicit. However, it'd be more verbose… to pass the test profile.

So, the current approach is sensible.

Could you write a comment that `prepareTestProfile` stores the profile to test in `testProfile`? It could be a javadoc for the `testProfile` field.

honkar-jdk replied:
> I'd probably return the resulting profile to run test on, it will make data flow explicit. However, it'd be more verbose… to pass the test profile.

My initial thought was same - return the profile, but chose to keep it as static var to reduce the number of arguments prepareTestProfile() & testProfile() and it keeps the test code simple.

> Could you write a comment that prepareTestProfile stores the profile to test in testProfile? It could be a javadoc for the testProfile field.

Updated.

aivanov-jdk commented on src/java.desktop/share/classes/java/awt/color/ICC_Profile.java:
```diff
@@ -107,15 +107,23 @@ public sealed class ICC_Profile implements Serializable
+
+    /**
+     * Set to {@code true} for {@code BuiltInProfile}, {@code false} otherwise.
+     * This check is used in {@link #setData(int, byte[])} to prevent modifying
```
```suggestion
     * This flag is used in {@link #setData(int, byte[])} to prevent modifying
```

aivanov-jdk commented on src/java.desktop/share/classes/java/awt/color/ICC_Profile.java:
```diff
@@ -107,15 +107,23 @@ public sealed class ICC_Profile implements Serializable
         /*
-         * Deferral is only used for standard profiles. Enabling the appropriate
-         * access privileges is handled at a lower level.
+         * ProfileDeferralInfo is used for only built-in profile creation and
+         * all built-in profiles should be constructed using it.
```
```suggestion
         * ProfileDeferralInfo is used for built-in profile creation only, and
         * all built-in profiles should be constructed using it.
```

aivanov-jdk commented on src/java.desktop/share/classes/java/awt/color/ICC_Profile.java:
```diff
@@ -763,14 +771,20 @@ private interface BuiltInProfile {
      * The ID will be 0 until the profile is loaded.
+     *
+     * <p>
+     * Note: {@code ProfileDeferralInfo} is used for only built-in profile
+     * creation and all built-in profiles should be constructed using it.
```
```suggestion
     * Note: {@code ProfileDeferralInfo} is used for built-in profile
     * creation only, and all built-in profiles should be constructed using it.
```

aivanov-jdk commented on test/jdk/java/awt/color/ICC_ProfileSetNullDataTest.java:
```diff
@@ -23,29 +23,33 @@
+            ICC_Profile profile = ICC_Profile.getInstance(builtInProfile.getData());
+            try {
+                profile.setData(ICC_Profile.icSigCmykData, null);
+                throw new RuntimeException("IAE expected, but not thrown for"
```
```suggestion
                throw new RuntimeException("IAE expected, but not thrown for "
```

aivanov-jdk commented on test/jdk/java/awt/color/ICC_Profile/BuiltInProfileCheck/BuiltInProfileCheck.java:
```diff
@@ -0,0 +1,169 @@
+            try {
+                // Try updating a built-in profile, IAE is expected
+                testProfile.setData(HEADER_TAG, headerData);
+                throw new RuntimeException("Test Failed! IAE NOT thrown for profile"
```
```suggestion
                throw new RuntimeException("Test Failed! IAE NOT thrown for profile "
```

aivanov-jdk reviewed (approved):
Looks good to me, yet there are a few small typos, fixes…

honkar-jdk commented:
/integrate
